sky/utils/policy_utils.py

```diff
@@ -0,0 +1,162 @@
+                'interface. Please check with your policy admin for details.')
+    return policy_cls
+
+
+@typing.overload
+def apply(
+    entrypoint: Union['dag_lib.Dag', 'task_lib.Task'],
+    apply_skypilot_config: Literal[True] = True,
+    operation_args: Optional[policy_lib.OperationArgs] = None,
+) -> 'dag_lib.Dag':
+    ...
+
+
+@typing.overload
+def apply(
+    entrypoint: Union['dag_lib.Dag', 'task_lib.Task'],
+    apply_skypilot_config: Literal[False],
+    operation_args: Optional[policy_lib.OperationArgs] = None,
+) -> Tuple['dag_lib.Dag', skypilot_config.NestedConfig]:
+    ...
+
+
+def apply(
+    entrypoint: Union['dag_lib.Dag', 'task_lib.Task'],
+    apply_skypilot_config: bool = True,
+    operation_args: Optional[policy_lib.OperationArgs] = None,
+) -> Union['dag_lib.Dag', Tuple['dag_lib.Dag', skypilot_config.NestedConfig]]:
+    """Applies user-defined policy to a DAG or a task.
+
+    It mutates a Dag by applying user-defined policy and also updates the
+    global SkyPilot config if there is any changes made by the policy.
+
+    Args:
+        dag: The dag to be mutated by the policy.
+        apply_skypilot_config: Whether to apply the skypilot config changes to
+            the global skypilot config.
+
+    Returns:
+        The mutated dag or task.
+        Or, a tuple of the mutated dag and path to the mutated skypilot
+        config, if apply_skypilot_config is set to False.
+    """
+    if isinstance(entrypoint, task_lib.Task):
+        dag = dag_lib.Dag()
+        dag.add(entrypoint)
+    else:
+        dag = entrypoint
+
+    policy = skypilot_config.get_nested(('admin_policy',), None)
+    policy_cls = _get_policy_cls(policy)
+    if policy_cls is None:
+        if apply_skypilot_config:
+            return dag
+        else:
+            return dag, skypilot_config.to_dict()
+
+    logger.info(f'Applying policy: {policy}')
+    original_config = skypilot_config.to_dict()
+    config = copy.deepcopy(original_config)
+    mutated_dag = dag_lib.Dag()
+    mutated_dag.name = dag.name
+
+    mutated_config = None
+    for task in dag.tasks:
+        user_request = policy_lib.UserRequest(task, config, operation_args)
+        try:
+            mutated_user_request = policy_cls.validate_and_mutate(user_request)
+        except Exception as e:  # pylint: disable=broad-except
+            with ux_utils.print_exception_no_traceback():
+                raise exceptions.UserRequestRejectedByPolicy(
+                    f'{colorama.Fore.RED}User request rejected by policy '
+                    f'{policy!r}{colorama.Fore.RESET}: '
+                    f'{common_utils.format_exception(e, use_bracket=True)}'
+                ) from e
+        if mutated_config is None:
+            mutated_config = mutated_user_request.skypilot_config
+        else:
+            if mutated_config != mutated_user_request.skypilot_config:
+                with ux_utils.print_exception_no_traceback():
```

Comment:
When would this happen in practice? In a pipeline & the config change is somehow dynamic?  E.g., "task1 -> task2", and a cluster list injection changes in between?

Reply:
My comment was a question :) If you meant "this will never happen for now", can we add a comment?

Reply:
Added a comment. Thanks!